_Lambda School Data Science_

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

## Download data

We’ll work with a dataset of [3 Million Instacart Orders, Open Sourced](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2)!

In [1]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2019-04-29 09:15:54--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.136.37
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.136.37|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M   750KB/s    in 4m 10s  

2019-04-29 09:20:05 (802 KB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [2]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

x instacart_2017_05_01/
x instacart_2017_05_01/._aisles.csv
x instacart_2017_05_01/aisles.csv
x instacart_2017_05_01/._departments.csv
x instacart_2017_05_01/departments.csv
x instacart_2017_05_01/._order_products__prior.csv
x instacart_2017_05_01/order_products__prior.csv
x instacart_2017_05_01/._order_products__train.csv
x instacart_2017_05_01/order_products__train.csv
x instacart_2017_05_01/._orders.csv
x instacart_2017_05_01/orders.csv
x instacart_2017_05_01/._products.csv
x instacart_2017_05_01/products.csv


In [1]:
%cd instacart_2017_05_01

/Users/ridleyleisy/Documents/lambda/unit_one/DS-Unit-1-Sprint-2-Data-Wrangling-and-Storytelling/module1-join-and-reshape-data/instacart_2017_05_01


In [2]:
!ls -lh *.csv

-rw-r--r--@ 1 ridleyleisy  staff   2.5K May  2  2017 aisles.csv
-rw-r--r--@ 1 ridleyleisy  staff   270B May  2  2017 departments.csv
-rw-r--r--@ 1 ridleyleisy  staff   551M May  2  2017 order_products__prior.csv
-rw-r--r--@ 1 ridleyleisy  staff    24M May  2  2017 order_products__train.csv
-rw-r--r--@ 1 ridleyleisy  staff   104M May  2  2017 orders.csv
-rw-r--r--@ 1 ridleyleisy  staff   2.1M May  2  2017 products.csv


# Join Datasets

## Goal: Reproduce this example

The first two orders for user id 1:

In [3]:
from IPython.display import display, Image
url = 'https://cdn-images-1.medium.com/max/1600/1*vYGFQCafJtGBBX5mbl0xyw.png'
example = Image(url=url, width=600)

display(example)

## Load data

Here's a list of all six CSV filenames

In [4]:
!ls -lh *.csv

-rw-r--r--@ 1 ridleyleisy  staff   2.5K May  2  2017 aisles.csv
-rw-r--r--@ 1 ridleyleisy  staff   270B May  2  2017 departments.csv
-rw-r--r--@ 1 ridleyleisy  staff   551M May  2  2017 order_products__prior.csv
-rw-r--r--@ 1 ridleyleisy  staff    24M May  2  2017 order_products__train.csv
-rw-r--r--@ 1 ridleyleisy  staff   104M May  2  2017 orders.csv
-rw-r--r--@ 1 ridleyleisy  staff   2.1M May  2  2017 products.csv


For each CSV
- Load it with pandas
- Look at the dataframe's shape
- Look at its head (first rows)
- `display(example)`
- Which columns does it have in common with the example we want to reproduce?

### aisles

In [5]:
import pandas as pd

In [6]:
path = '/Users/ridleyleisy/Documents/lambda/unit_one/DS-Unit-1-Sprint-2-Data-Wrangling-and-Storytelling/module1-join-and-reshape-data/instacart_2017_05_01/'

In [7]:
!head aisles.csv

aisle_id,aisle
1,prepared soups salads
2,specialty cheeses
3,energy granola bars
4,instant foods
5,marinades meat preparation
6,other
7,packaged meat
8,bakery desserts
9,pasta sauce


In [8]:
!wc aisles.csv

     135     311    2603 aisles.csv


In [9]:
aisles = pd.read_csv(path + 'aisles.csv')

In [10]:
aisles

aisle_id                          aisle
0           1          prepared soups salads
1           2              specialty cheeses
2           3            energy granola bars
3           4                  instant foods
4           5     marinades meat preparation
5           6                          other
6           7                  packaged meat
7           8                bakery desserts
8           9                    pasta sauce
9          10               kitchen supplies
10         11               cold flu allergy
11         12                    fresh pasta
12         13                 prepared meals
13         14         tofu meat alternatives
14         15               packaged seafood
15         16                    fresh herbs
16         17             baking ingredients
17         18   bulk dried fruits vegetables
18         19                  oils vinegars
19         20                   oral hygiene
20         21                packaged cheese
21         22                      hair care
22         23                  popcorn jerky
23         24                   fresh fruits
24         25                           soap
25         26                         coffee
26         27                  beers coolers
27         28                      red wines
28         29          honeys syrups nectars
29         30                   latino foods
..        ...                            ...
104       105     doughs gelatins bake mixes
105       106         hot dogs bacon sausage
106       107                 chips pretzels
107       108           other creams cheeses
108       109                      skin care
109       110           pickled goods olives
110       111     plates bowls cups flatware
111       112                          bread
112       113                   frozen juice
113       114              cleaning products
114       115  water seltzer sparkling water
115       116                 frozen produce
116       117         nuts seeds dried fruit
117       118                      first aid
118       119                 frozen dessert
119       120                         yogurt
120       121                         cereal
121       122                   meat counter
122       123     packaged vegetables fruits
123       124                        spirits
124       125            trail mix snack mix
125       126                  feminine care
126       127              body lotions soap
127       128           tortillas flat bread
128       129        frozen appetizers sides
129       130       hot cereal pancake mixes
130       131                      dry pasta
131       132                         beauty
132       133     muscles joints pain relief
133       134     specialty wines champagnes

[134 rows x 2 columns]

### departments

In [11]:
departments = pd.read_csv(path + 'departments.csv')

### order_products__prior

In [12]:
!head order_products__prior.csv

order_id,product_id,add_to_cart_order,reordered
2,33120,1,1
2,28985,2,1
2,9327,3,0
2,45918,4,1
2,30035,5,0
2,17794,6,1
2,40141,7,1
2,1819,8,1
2,43668,9,0


In [13]:
prior_order = pd.read_csv(path + 'order_products__prior.csv')

In [14]:
prior_order

order_id  product_id  add_to_cart_order  reordered
0                2       33120                  1          1
1                2       28985                  2          1
2                2        9327                  3          0
3                2       45918                  4          1
4                2       30035                  5          0
5                2       17794                  6          1
6                2       40141                  7          1
7                2        1819                  8          1
8                2       43668                  9          0
9                3       33754                  1          1
10               3       24838                  2          1
11               3       17704                  3          1
12               3       21903                  4          1
13               3       17668                  5          1
14               3       46667                  6          1
15               3       17461                  7          1
16               3       32665                  8          1
17               4       46842                  1          0
18               4       26434                  2          1
19               4       39758                  3          1
20               4       27761                  4          1
21               4       10054                  5          1
22               4       21351                  6          1
23               4       22598                  7          1
24               4       34862                  8          1
25               4       40285                  9          1
26               4       17616                 10          1
27               4       25146                 11          1
28               4       32645                 12          1
29               4       41276                 13          1
...            ...         ...                ...        ...
32434459   3421080       41950                  4          0
32434460   3421080       31717                  5          0
32434461   3421080       12935                  6          1
32434462   3421080       25122                  7          0
32434463   3421080       10667                  8          0
32434464   3421080       38061                  9          0
32434465   3421081       38185                  1          0
32434466   3421081       12218                  2          0
32434467   3421081       32299                  3          0
32434468   3421081        3060                  4          0
32434469   3421081       20539                  5          0
32434470   3421081       35221                  6          0
32434471   3421081       12861                  7          0
32434472   3421082       17279                  1          1
32434473   3421082       12738                  2          1
32434474   3421082       16797                  3          0
32434475   3421082       43352                  4          1
32434476   3421082       32700                  5          1
32434477   3421082       12023                  6          0
32434478   3421082       47941                  7          0
32434479   3421083        7854                  1          0
32434480   3421083       45309                  2          0
32434481   3421083       21162                  3          0
32434482   3421083       18176                  4          1
32434483   3421083       35211                  5          0
32434484   3421083       39678                  6          1
32434485   3421083       11352                  7          0
32434486   3421083        4600                  8          0
32434487   3421083       24852                  9          1
32434488   3421083        5020                 10          1

[32434489 rows x 4 columns]

In [15]:
prior_order.groupby(['order_id'])['product_id'].count().mean()

10.088883421247614

### order_products__train

In [16]:
# no difference in train vs prior

In [17]:
train_order = pd.read_csv(path + 'order_products__train.csv')

### orders

In [18]:
orders = pd.read_csv(path + 'orders.csv')

In [19]:
orders.shape

(3421083, 7)

In [20]:
orders.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [21]:
orders.isna().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### products

In [22]:
products = pd.read_csv(path + 'products.csv')

In [23]:
products.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  
0             19  
1             13  
2              7  
3              1  
4             13

## Concatenate order_products__prior and order_products__train

In [24]:
order_products = pd.concat([train_order,prior_order])

In [25]:
order_products

order_id  product_id  add_to_cart_order  reordered
0                1       49302                  1          1
1                1       11109                  2          1
2                1       10246                  3          0
3                1       49683                  4          0
4                1       43633                  5          1
5                1       13176                  6          0
6                1       47209                  7          0
7                1       22035                  8          1
8               36       39612                  1          0
9               36       19660                  2          1
10              36       49235                  3          0
11              36       43086                  4          1
12              36       46620                  5          1
13              36       34497                  6          1
14              36       48679                  7          1
15              36       46979                  8          1
16              38       11913                  1          0
17              38       18159                  2          0
18              38        4461                  3          0
19              38       21616                  4          1
20              38       23622                  5          0
21              38       32433                  6          0
22              38       28842                  7          0
23              38       42625                  8          0
24              38       39693                  9          0
25              96       20574                  1          1
26              96       30391                  2          0
27              96       40706                  3          1
28              96       25610                  4          0
29              96       27966                  5          1
...            ...         ...                ...        ...
32434459   3421080       41950                  4          0
32434460   3421080       31717                  5          0
32434461   3421080       12935                  6          1
32434462   3421080       25122                  7          0
32434463   3421080       10667                  8          0
32434464   3421080       38061                  9          0
32434465   3421081       38185                  1          0
32434466   3421081       12218                  2          0
32434467   3421081       32299                  3          0
32434468   3421081        3060                  4          0
32434469   3421081       20539                  5          0
32434470   3421081       35221                  6          0
32434471   3421081       12861                  7          0
32434472   3421082       17279                  1          1
32434473   3421082       12738                  2          1
32434474   3421082       16797                  3          0
32434475   3421082       43352                  4          1
32434476   3421082       32700                  5          1
32434477   3421082       12023                  6          0
32434478   3421082       47941                  7          0
32434479   3421083        7854                  1          0
32434480   3421083       45309                  2          0
32434481   3421083       21162                  3          0
32434482   3421083       18176                  4          1
32434483   3421083       35211                  5          0
32434484   3421083       39678                  6          1
32434485   3421083       11352                  7          0
32434486   3421083        4600                  8          0
32434487   3421083       24852                  9          1
32434488   3421083        5020                 10          1

[33819106 rows x 4 columns]

## Get a subset of orders — the first two orders for user id 1

From `orders` dataframe:
- user_id
- order_id
- order_number
- order_dow
- order_hour_of_day

In [26]:
subset = orders[(orders['user_id'] == 1) & (orders['order_number'] <= 2)][['user_id','order_id','order_number','order_dow','order_hour_of_day']]

In [27]:
subset

user_id  order_id  order_number  order_dow  order_hour_of_day
0        1   2539329             1          2                  8
1        1   2398795             2          3                  7

## Merge dataframes

Merge the subset from `orders` with columns from `order_products`

In [28]:
columns = ['order_id','add_to_cart_order','product_id']

In [29]:
order_products[columns].head()

order_id  add_to_cart_order  product_id
0         1                  1       49302
1         1                  2       11109
2         1                  3       10246
3         1                  4       49683
4         1                  5       43633

Merge with columns from `products`

In [30]:
merged = pd.merge(subset, order_products[columns],how='inner',on='order_id')

In [31]:
merged

user_id  order_id  order_number  order_dow  order_hour_of_day  \
0         1   2539329             1          2                  8   
1         1   2539329             1          2                  8   
2         1   2539329             1          2                  8   
3         1   2539329             1          2                  8   
4         1   2539329             1          2                  8   
5         1   2398795             2          3                  7   
6         1   2398795             2          3                  7   
7         1   2398795             2          3                  7   
8         1   2398795             2          3                  7   
9         1   2398795             2          3                  7   
10        1   2398795             2          3                  7   

    add_to_cart_order  product_id  
0                   1         196  
1                   2       14084  
2                   3       12427  
3                   4       26088  
4                   5       26405  
5                   1         196  
6                   2       10258  
7                   3       12427  
8                   4       13176  
9                   5       26088  
10                  6       13032

In [32]:
products

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
5               6                                       Dry Nose Oil   
6               7                     Pure Coconut Water With Orange   
7               8                  Cut Russet Potatoes Steam N' Mash   
8               9                  Light Strawberry Blueberry Yogurt   
9              10     Sparkling Orange Juice & Prickly Pear Beverage   
10             11                                  Peach Mango Juice   
11             12                         Chocolate Fudge Layer Cake   
12             13                                  Saline Nasal Mist   
13             14                     Fresh Scent Dishwasher Cleaner   
14             15                           Overnight Diapers Size 6   
15             16                      Mint Chocolate Flavored Syrup   
16             17                                  Rendered Duck Fat   
17             18                Pizza for One Suprema  Frozen Pizza   
18             19   Gluten Free Quinoa Three Cheese & Mushroom Blend   
19             20     Pomegranate Cranberry & Aloe Vera Enrich Drink   
20             21                   Small & Medium Dental Dog Treats   
21             22                  Fresh Breath Oral Rinse Mild Mint   
22             23                             Organic Turkey Burgers   
23             24  Tri-Vi-Sol® Vitamins A-C-and D Supplement Drop...   
24             25            Salted Caramel Lean Protein & Fiber Bar   
25             26        Fancy Feast Trout Feast Flaked Wet Cat Food   
26             27  Complete Spring Water Foaming Antibacterial Ha...   
27             28                                  Wheat Chex Cereal   
28             29  Fresh Cut Golden Sweet No Salt Added Whole Ker...   
29             30         Three Cheese Ziti, Marinara with Meatballs   
...           ...                                                ...   
49658       49659                            Organic Creamed Coconut   
49659       49660                        Professionals Sleek Shampoo   
49660       49661                                              Porto   
49661       49662                       Bacon Cheddar Pretzel Pieces   
49662       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   
49663       49664                     Lemon Cayenne Drinking Vinegar   
49664       49665      Super Dark Coconut Ash & Banana Chocolate Bar   
49665       49666                      Ginger Snaps Snacking Cookies   
49666       49667           Enchilada with Spanish Rice & Beans Meal   
49667       49668                     Apple Cinnamon Scented Candles   
49668       49669                                   K Cup Dark Blend   
49669       49670                                Beef Summer Sausage   
49670       49671                               Milk Chocolate Drops   
49671       49672                             Cafe Mocha K-Cup Packs   
49672       49673              Stone Baked Multi Grain Artisan Rolls   
49673       49674            Frozen Greek Yogurt Bars Chocolate Chip   
49674       49675                Cinnamon Dolce Keurig Brewed K Cups   
49675       49676                             Ultra Red Energy Drink   
49676       49677                    Thick & Chunky Sloppy Joe Sauce   
49677       49678                    Large Chicken & Cheese Taquitos   
49678       49679                            Famous Chocolate Wafers   
49679       49680                 All Natural Creamy Caesar Dressing   
49680       49681            Spaghetti with Meatballs and Sauce Meal   
49681       49682                              

In [33]:
final = pd.merge(merged, products[['product_id','product_name']],how='inner',on='product_id')

In [34]:
final = final.sort_values(by=['order_id','add_to_cart_order'])

In [35]:
display(final)

user_id  order_id  order_number  order_dow  order_hour_of_day  \
1         1   2398795             2          3                  7   
8         1   2398795             2          3                  7   
4         1   2398795             2          3                  7   
9         1   2398795             2          3                  7   
6         1   2398795             2          3                  7   
10        1   2398795             2          3                  7   
0         1   2539329             1          2                  8   
2         1   2539329             1          2                  8   
3         1   2539329             1          2                  8   
5         1   2539329             1          2                  8   
7         1   2539329             1          2                  8   

    add_to_cart_order  product_id                             product_name  
1                   1         196                                     Soda  
8                   2       10258                               Pistachios  
4                   3       12427                      Original Beef Jerky  
9                   4       13176                   Bag of Organic Bananas  
6                   5       26088               Aged White Cheddar Popcorn  
10                  6       13032                    Cinnamon Toast Crunch  
0                   1         196                                     Soda  
2                   2       14084  Organic Unsweetened Vanilla Almond Milk  
3                   3       12427                      Original Beef Jerky  
5                   4       26088               Aged White Cheddar Popcorn  
7                   5       26405         XL Pick-A-Size Paper Towel Rolls

# Reshape Datasets

## Why reshape data?

#### Some libraries prefer data in different formats

For example, the Seaborn data visualization library prefers data in "Tidy" format often (but not always).

> "[Seaborn will be most powerful when your datasets have a particular organization.](https://seaborn.pydata.org/introduction.html#organizing-datasets) This format ia alternately called “long-form” or “tidy” data and is described in detail by Hadley Wickham. The rules can be simply stated:

> - Each variable is a column
- Each observation is a row

> A helpful mindset for determining whether your data are tidy is to think backwards from the plot you want to draw. From this perspective, a “variable” is something that will be assigned a role in the plot."

#### Data science is often about putting square pegs in round holes

Here's an inspiring [video clip from _Apollo 13_](https://www.youtube.com/watch?v=ry55--J4_VQ): “Invent a way to put a square peg in a round hole.” It's a good metaphor for data wrangling!

## Hadley Wickham's Examples

From his paper, [Tidy Data](http://vita.had.co.nz/papers/tidy-data.html)

In [36]:
%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

"Table 1 provides some data about an imaginary experiment in a format commonly seen in the wild. 

The table has two columns and three rows, and both rows and columns are labelled."

In [37]:
table1

treatmenta  treatmentb
John Smith           NaN           2
Jane Doe            16.0          11
Mary Johnson         3.0           1

"There are many ways to structure the same underlying data. 

Table 2 shows the same data as Table 1, but the rows and columns have been transposed. The data is the same, but the layout is different."

In [38]:
table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

"Table 3 reorganises Table 1 to make the values, variables and obserations more clear.

Table 3 is the tidy version of Table 1. Each row represents an observation, the result of one treatment on one person, and each column is a variable."

| name         | trt | result |
|--------------|-----|--------|
| John Smith   | a   | -      |
| Jane Doe     | a   | 16     |
| Mary Johnson | a   | 3      |
| John Smith   | b   | 2      |
| Jane Doe     | b   | 11     |
| Mary Johnson | b   | 1      |

## Table 1 --> Tidy

We can use the pandas `melt` function to reshape Table 1 into Tidy format.

In [39]:
help(pd.melt)

Help on function melt in module pandas.core.reshape.melt:

melt(frame, id_vars=None, value_vars=None, var_name=None, value_name='value', col_level=None)
    Unpivots a DataFrame from wide format to long format, optionally
    leaving identifier variables set.
    
    This function is useful to massage a DataFrame into a format where one
    or more columns are identifier variables (`id_vars`), while all other
    columns, considered measured variables (`value_vars`), are "unpivoted" to
    the row axis, leaving just two non-identifier columns, 'variable' and
    'value'.
    
    
    Parameters
    ----------
    frame : DataFrame
    id_vars : tuple, list, or ndarray, optional
        Column(s) to use as identifier variables.
    value_vars : tuple, list, or ndarray, optional
        Column(s) to unpivot. If not specified, uses all columns that
        are not set as `id_vars`.
    var_name : scalar
        Name to use for the 'variable' column. If None it uses
        ``frame.colum

## Table 2 --> Tidy

In [40]:
##### LEAVE BLANK --an assignment exercise #####

## Tidy --> Table 1

The `pivot_table` function is the inverse of `melt`.

## Tidy --> Table 2

In [41]:
##### LEAVE BLANK --an assignment exercise #####

# Seaborn example

The rules can be simply stated:

- Each variable is a column
- Each observation is a row

A helpful mindset for determining whether your data are tidy is to think backwards from the plot you want to draw. From this perspective, a “variable” is something that will be assigned a role in the plot."

In [42]:
sns.catplot(x='trt', y='result', col='name', 
            kind='bar', data=tidy, height=2);

NameError: name 'tidy' is not defined

## Now with Instacart data

In [ ]:
products = pd.read_csv('products.csv')

order_products = pd.concat([pd.read_csv('order_products__prior.csv'), 
                            pd.read_csv('order_products__train.csv')])

orders = pd.read_csv('orders.csv')

## Goal: Reproduce part of this example

Instead of a plot with 50 products, we'll just do two — the first products from each list
- Half And Half Ultra Pasteurized
- Half Baked Frozen Yogurt

In [ ]:
from IPython.display import display, Image
url = 'https://cdn-images-1.medium.com/max/1600/1*wKfV6OV-_1Ipwrl7AjjSuw.png'
example = Image(url=url, width=600)

display(example)

So, given a `product_name` we need to calculate its `order_hour_of_day` pattern.

## Subset and Merge

One challenge of performing a merge on this data is that the `products` and `orders` datasets do not have any common columns that we can merge on. Due to this we will have to use the `order_products` dataset to provide the columns that we will use to perform the merge.

## 4 ways to reshape and plot

### 1. value_counts

### 2. crosstab

### 3. Pivot Table

### 4. melt

# Assignment

## Join Data Section

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

## Reshape Data Section

- Replicate the lesson code
- Complete the code cells we skipped near the beginning of the notebook
- Table 2 --> Tidy
- Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [48]:
product_list = ['Banana','Bag of Organic Bananas','Organic Strawberries','Organic Baby Spinach',
           'Organic Hass Avocado','Organic Avocado','Large Lemon','Strawberries','Limes','Organic Whole Milk']

In [49]:
products = pd.read_csv(path + 'products.csv')

In [50]:
products

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
5               6                                       Dry Nose Oil   
6               7                     Pure Coconut Water With Orange   
7               8                  Cut Russet Potatoes Steam N' Mash   
8               9                  Light Strawberry Blueberry Yogurt   
9              10     Sparkling Orange Juice & Prickly Pear Beverage   
10             11                                  Peach Mango Juice   
11             12                         Chocolate Fudge Layer Cake   
12             13                                  Saline Nasal Mist   
13             14                     Fresh Scent Dishwasher Cleaner   
14             15                           Overnight Diapers Size 6   
15             16                      Mint Chocolate Flavored Syrup   
16             17                                  Rendered Duck Fat   
17             18                Pizza for One Suprema  Frozen Pizza   
18             19   Gluten Free Quinoa Three Cheese & Mushroom Blend   
19             20     Pomegranate Cranberry & Aloe Vera Enrich Drink   
20             21                   Small & Medium Dental Dog Treats   
21             22                  Fresh Breath Oral Rinse Mild Mint   
22             23                             Organic Turkey Burgers   
23             24  Tri-Vi-Sol® Vitamins A-C-and D Supplement Drop...   
24             25            Salted Caramel Lean Protein & Fiber Bar   
25             26        Fancy Feast Trout Feast Flaked Wet Cat Food   
26             27  Complete Spring Water Foaming Antibacterial Ha...   
27             28                                  Wheat Chex Cereal   
28             29  Fresh Cut Golden Sweet No Salt Added Whole Ker...   
29             30         Three Cheese Ziti, Marinara with Meatballs   
...           ...                                                ...   
49658       49659                            Organic Creamed Coconut   
49659       49660                        Professionals Sleek Shampoo   
49660       49661                                              Porto   
49661       49662                       Bacon Cheddar Pretzel Pieces   
49662       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   
49663       49664                     Lemon Cayenne Drinking Vinegar   
49664       49665      Super Dark Coconut Ash & Banana Chocolate Bar   
49665       49666                      Ginger Snaps Snacking Cookies   
49666       49667           Enchilada with Spanish Rice & Beans Meal   
49667       49668                     Apple Cinnamon Scented Candles   
49668       49669                                   K Cup Dark Blend   
49669       49670                                Beef Summer Sausage   
49670       49671                               Milk Chocolate Drops   
49671       49672                             Cafe Mocha K-Cup Packs   
49672       49673              Stone Baked Multi Grain Artisan Rolls   
49673       49674            Frozen Greek Yogurt Bars Chocolate Chip   
49674       49675                Cinnamon Dolce Keurig Brewed K Cups   
49675       49676                             Ultra Red Energy Drink   
49676       49677                    Thick & Chunky Sloppy Joe Sauce   
49677       49678                    Large Chicken & Cheese Taquitos   
49678       49679                            Famous Chocolate Wafers   
49679       49680                 All Natural Creamy Caesar Dressing   
49680       49681            Spaghetti with Meatballs and Sauce Meal   
49681       49682                              

In [45]:
order_products.head()

order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

In [84]:
order_products

order_id  product_id  add_to_cart_order  reordered
0                1       49302                  1          1
1                1       11109                  2          1
2                1       10246                  3          0
3                1       49683                  4          0
4                1       43633                  5          1
5                1       13176                  6          0
6                1       47209                  7          0
7                1       22035                  8          1
8               36       39612                  1          0
9               36       19660                  2          1
10              36       49235                  3          0
11              36       43086                  4          1
12              36       46620                  5          1
13              36       34497                  6          1
14              36       48679                  7          1
15              36       46979                  8          1
16              38       11913                  1          0
17              38       18159                  2          0
18              38        4461                  3          0
19              38       21616                  4          1
20              38       23622                  5          0
21              38       32433                  6          0
22              38       28842                  7          0
23              38       42625                  8          0
24              38       39693                  9          0
25              96       20574                  1          1
26              96       30391                  2          0
27              96       40706                  3          1
28              96       25610                  4          0
29              96       27966                  5          1
...            ...         ...                ...        ...
32434459   3421080       41950                  4          0
32434460   3421080       31717                  5          0
32434461   3421080       12935                  6          1
32434462   3421080       25122                  7          0
32434463   3421080       10667                  8          0
32434464   3421080       38061                  9          0
32434465   3421081       38185                  1          0
32434466   3421081       12218                  2          0
32434467   3421081       32299                  3          0
32434468   3421081        3060                  4          0
32434469   3421081       20539                  5          0
32434470   3421081       35221                  6          0
32434471   3421081       12861                  7          0
32434472   3421082       17279                  1          1
32434473   3421082       12738                  2          1
32434474   3421082       16797                  3          0
32434475   3421082       43352                  4          1
32434476   3421082       32700                  5          1
32434477   3421082       12023                  6          0
32434478   3421082       47941                  7          0
32434479   3421083        7854                  1          0
32434480   3421083       45309                  2          0
32434481   3421083       21162                  3          0
32434482   3421083       18176                  4          1
32434483   3421083       35211                  5          0
32434484   3421083       39678                  6          1
32434485   3421083       11352                  7          0
32434486   3421083        4600                  8          0
32434487   3421083       24852                  9          1
32434488   3421083        5020                 10          1

[33819106 rows x 4 columns]

In [0]:
flights = sns.load_dataset('flights')

In [0]:
##### YOUR CODE HERE #####

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)